### ISSUE TITLE
Error when trying to build `java_binary` on Windows

### ISSUE THREAD

**Issue by oferb** [bazelbuild/bazel#5260]
Error when trying to build `java_binary` on Windows

### Feature requests: what underlying problem are you trying to solve with this feature?

Compile :)

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.
Clone `https://github.com/google/startup-os`
Run `bazel build //tools/reviewer/service`

This results in the following (run with `--verbose_failures`):
```
bazel build //tools/reviewer/service --verbose_failures
ERROR: C:/root/storage/nodejs/startup-os/tools/reviewer/service/BUILD:27:1: Building tools/reviewer/service/libservice.jar (1 source file) and running annotation processors (AutoFactoryProcessor) failed (Exit 1): java.exe failed: error executing command
  cd C:/users/locked/_bazel_locked/zrv4ye2p/execroot/__main__
  SET LC_CTYPE=en_US.UTF-8
  external/local_jdk/bin/java.exe -Xbootclasspath/p:external/bazel_tools/third_party/java/jdk/langtools/javac-9+181-r4173-1.jar -jar external/bazel_tools/tools/jdk/JavaBuilder_deploy.jar @bazel-out/x64_windows-fastbuild/bin/tools/reviewer/service/libservice.jar-2.params
tools\reviewer\service\CodeReviewService.java:17: error: cannot access com.google.startupos.tools.reviewer.service
package com.google.startupos.tools.reviewer.service;
^
  /com/google/startupos/common/Time.class
Target //tools/reviewer/service:service failed to build
```

### What operating system are you running Bazel on?
Windows 10

### What's the output of `bazel info release`?
Release 0.13.0

### What's the output of `git remote get-url origin ; git rev-parse master ; git rev-parse HEAD` ?
`git remote get-url origin`: git@github.com:google/startup-os.git
`git rev-parse master`: 2c3bcb49f8d1092241fd982874dce621bbcde91f
`git rev-parse HEAD`:    2c3bcb49f8d1092241fd982874dce621bbcde91f

###  Have you found anything relevant by searching the web?
Nope

### Any other information, logs, or outputs that you want to share?
Seems to be an issue with `java_binary`, not `java_library`:
8 different `java_library` targets successfully compiled.
2 different `java_binary` targets resulted in the error.

This same error happened on 2 different Windows machines, and successfully compiled on 5 Ubuntu & macOS ones.

**Comment by laszlocsomor** [comment]
I distilled a minimal repro case, see https://github.com/laszlocsomor/projects/commit/c0d5a9cbaf5a406537fe4b3fbd2287c67d5f3067.

`bazel build //foo --verbose_failures` succeeds on Linux, but fails on Windows the same way as it does for @oferb . Culprit is the `java_plugin.generates_api = True` BUILD rule attribute. That's all I know fow now.

**Comment by laszlocsomor** [comment]
Found the culprit: the Java header compilation creates header jars with zip entries that use backslashes as path separators instead of forward slashes. This confuses javac or the annotation processor.

I verified this as the culprit by copying the `bazel-bin/foo/libbar-hjar.jar` from Linux to Windows and retried the failing command, and it now works. Also, `unzip -l` shows the zip entries as different, but the files in the jar are the same:

```
# Bad header jar, generated on Windows:
$ unzip -l ../../../bazel-bin/foo/libbar-hjar.jar
Archive:  ../../../bazel-bin/foo/libbar-hjar.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
       93  2010-01-01 00:00   com\google\bar\Bar.class
        0  2010-01-01 00:00   META-INF/
       69  2010-01-01 00:00   META-INF/MANIFEST.MF
---------                     -------
      162                     3 files

# Good header jar, generated on Linux
$ unzip -l ../libbar-hjar.jar
Archive:  ../libbar-hjar.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
       93  2010-01-01 00:00   com/google/bar/Bar.class
        0  2010-01-01 00:00   META-INF/
       69  2010-01-01 00:00   META-INF/MANIFEST.MF
---------                     -------
      162                     3 files

# Diff the extracted files
$ diff com/google/bar/Bar.class ../good/com/google/bar/Bar.class

$ diff META-INF/MANIFEST.MF ../good/META-INF/MANIFEST.MF
```

**Comment by laszlocsomor** [comment]
Summary: Turbine writes zip entries into the hjar using backslashes on Windows:

```
C:\_bazel\c39O6PEq\execroot\__main__>external\local_jdk\bin\java.exe -Xverify:none -Xbootclasspath/p:external/bazel_tools/third_party/java/jdk/langtools/javac-9+181-r4173-1.jar -jar external/bazel_tools/tools/jdk/turbine_deploy.jar --output c:/tmp/foo-hjar.jar --temp_dir c:/tmp --bootclasspath external/local_jdk/jre/lib/rt.jar external/local_jdk/jre/lib/resources.jar external/local_jdk/jre/lib/jsse.jar external/local_jdk/jre/lib/jce.jar external/local_jdk/jre/lib/charsets.jar external/local_jdk/jre/lib/ext/access-bridge-64.jar external/local_jdk/jre/lib/ext/cldrdata.jar external/local_jdk/jre/lib/ext/dnsns.jar external/local_jdk/jre/lib/ext/jaccess.jar external/local_jdk/jre/lib/ext/localedata.jar external/local_jdk/jre/lib/ext/nashorn.jar external/local_jdk/jre/lib/ext/sunec.jar external/local_jdk/jre/lib/ext/sunjce_provider.jar external/local_jdk/jre/lib/ext/sunmscapi.jar external/local_jdk/jre/lib/ext/sunpkcs11.jar external/local_jdk/jre/lib/ext/zipfs.jar --sources foo/Bar.java --javacopts -source 8 -target 8 -encoding UTF-8 -XDskipDuplicateBridges=true -g -parameters -- --target_label //foo:bar --processors common.AnnotProc --processorpath bazel-out/host/bin/foo/libannotproc.jar

C:\_bazel\c39O6PEq\execroot\__main__>c:\tools\msys64\usr\bin\unzip.exe -l c:\tmp\foo-hjar.jar
Archive:  c:\tmp\foo-hjar.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
       93  2010-01-01 00:00   com\google\bar\Bar.class
        0  2010-01-01 00:00   META-INF/
       69  2010-01-01 00:00   META-INF/MANIFEST.MF
---------                     -------
      162                     3 files
```

**Comment by laszlocsomor** [comment]
@cushon , @iirina : See my previous comment for a summary.

**Comment by dslomov** [comment]
Sounds like a release blocker to me?

**Comment by laszlocsomor** [comment]
Yes.

**Comment by laszlocsomor** [comment]
However it's not a regression, so it shouldn't block any ongoing release.

**Comment by laszlocsomor** [comment]
Actually, this isn't a release blocker as such, because it's been broken in every release.

**Comment by oferb** [comment]
Thanks for the prompt investigation! It's mind-boggling how many eng-hours the `/` vs `\` incompatibility has probably cost over the years.

I guess we have an interesting combination of technologies in StartupOS that surfaced this bug.
I think what we'll do once this is fixed, is build bazel from head and try it out again. This way we can have faster cycles.

**Comment by oferb** [comment]
Thanks! Perhaps you have a nightly build or something like that we could use, instead of building it ourselves?

**Comment by laszlocsomor** [comment]
Opening again, as it's a legitimate multiplatform build bug that we must fix.

@oferb : unfortunately we don't have one. But if your system is already set up to build with Bazel, you can easily build a Bazel binary from HEAD, just clone https://github.com/bazelbuild/bazel and `bazel --output_user_root=c:\tmp build //src:bazel.exe`

**Comment by cushon** [comment]
> Opening again, as it's a legitimate multiplatform build bug that we must fix.

Is it not fixed by 10141a3a351dd49376fc3e891634c5d91328a462?

**Comment by laszlocsomor** [comment]
Oh sorry, I didn't see that! Yes it is. Thanks!

**Comment by oferb** [comment]
Hey,
Trying to build from head on Windows. Running:
`bazel --output_user_root=c:\tmp build //src:bazel.exe`
Getting:
`ERROR: C:/root/storage/nodejs/bazel/src/BUILD:229:2: Executing genrule //src:package-zip failed (Exit 127) src/package-bazel.sh: line 49: zip: command not found`

Trying:
`bazel build //src:bazel1
Getting:
```

**Comment by laszlocsomor** [comment]
Ah, that's annoying. Culprit: #5265.
You can solve this by installing `zip`: open an MSYS shell and run `pacman -Syu zip unzip`. Follow the instructions MSYS gives you -- it may ask you to close its window and restart it.

**Comment by laszlocsomor** [comment]
I confirmed that startup-os builds successfully at https://github.com/google/startup-os/commit/c71236624c6f0de486a47970f3b3253fb26b2d47 using Bazel built at https://github.com/bazelbuild/bazel/commit/a5ecd8132d3742831eda9730614270604294e410.

It does NOT build at current startup-os HEAD (https://github.com/google/startup-os/commit/d58cd6f411fd0653fd1d6725646783b87e2ac050) but the reason looks like an error with startup-os' WORKSPACE file.

**Comment by oferb** [comment]
Is the fix in `0.14`?

**Comment by laszlocsomor** [comment]
@oferb : if you mean https://github.com/bazelbuild/bazel/commit/10141a3a351dd49376fc3e891634c5d91328a462, no it's not, because 0.14.0's baseline (5c3f5c9be7fa40d4fb3c35756891fab8483ca406) was merged on 2018/05/11 and the fix was merged on 2018/05/25, and was not cherry-picked into the release.
